src/theory/quantifiers/trigger.h

```diff
@@ -33,102 +32,180 @@ namespace inst {
 };
 
-/** A collect of nodes representing a trigger. */
-class Trigger {
- public:
-  ~Trigger();
+class HigherOrderTrigger;
 
+/** A collection of nodes representing a trigger. 
+*
+* This class is a wrapper around an underlying IMGenerator class, which implements various forms 
+* of matching. A trigger is a set of nodes with free variables that are used to guide instantiations
+* (for example, see "Efficient E-Matching for SMT Solvers" by de Moura et al). For example :
+*
+* quantified formula : forall x. P( x )
+*            trigger : P( x )
+*     ground context : ~P( a )
+*
+* Then E-matching matches P( x ) and P( a ), resulting in the match { x -> a } which is used to generate
+* the instantiation lemma :
+* forall x. P( x ) => P( a )
+*
+* We use a Trigger* t in a full effort check, where assume that t is associated with quantified formula q (see field d_f),
+* in the following way:
+*
+* t->resetInstantiationRound();      // setup initial information
+* t->reset( Node::null() );          // will produce instantiations based on matching with all terms
+* InstMatch baseMatch;
+* t->addInstantiations( baseMatch ); // add all instantiations based E-matching with this trigger and the current context
+*
+* This will result in (a set of) calls to d_quantEngine->addInstantiation( q, m1 )...d_quantEngine->addInstantiation( q, mn ),
+* where m1...mn are InstMatch objects. These calls add the corresponding instantiation lemma for (q,mi) on the output channel
+* associated with d_quantEngine.
+*
+* For more details, see functions below.
+*/
+class Trigger {
+  friend class IMGenerator;
+public:
+  virtual ~Trigger();
+  /** get the generator associated with this trigger */
   IMGenerator* getGenerator() { return d_mg; }
-
-  /** reset instantiation round (call this whenever equivalence
-   * classes have changed) */
+  /** Reset instantiation round. 
+  *
+  * Called once at beginning of an instantiation round.
+  */
   void resetInstantiationRound();
-  /** reset, eqc is the equivalence class to search in (search in any
-   * if eqc=null) */
+  /** Reset the trigger.
+  * eqc is the equivalence class to match ground terms, or any equivalence class if eqc is Node::null().
+  */
   void reset( Node eqc );
-  /** get next match.  must call reset( eqc ) once before this function. */
+  /** get next match.  Must call reset( eqc ) once before this function. */
   bool getNextMatch( Node f, InstMatch& m );
+  /** add all available instantiations, based on the current context
+  *
+  * This function makes the appropriate class to d_qe->addInstantiation(...) based on the current
+  * ground terms and equalities in the current context, via queries to functions in d_qe. Typically,
+  * this is implemented as a fixed point of getNextMatch within the underlying IMGenerator.
+  *
+  * baseMatch is a mapping of default values that should be used for variables that are not bound by this (not frequently used).
```

Comment:
What do you mean by not frequently used?  How do you not use it if it is being passed by reference?

Reply:
still have this question